### ISSUE TITLE
Function argument lines should not be put together to match print-width

### ISSUE THREAD

**Issue by MartinMuzatko** [prettier/prettier#3389]
**Prettier 1.8.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEA7gIYxgAWAFGMQDa0BGxYA1gJT7AA6U+f+qAM4A6EmXIAxHLTiUe-BfgDCAeQByEgJIBxLQBkAogAUAggBUAEgBp5ivsByw4AJwBudJPgCMABj8BfG147fGJBAE8sfGp0Zyt8AAdnOFcOAF4APk5bEP4ZGHxIKFQcAHN8NNCSHALSuBhKGGd0ODYc3MK6RmYWamgS0rbgu392of8QKxAIBNxoQWRQMJhkVDpBOCmGZx76gGUE5kdS5CaWqYATCDBV9c2QRw3nGCMd0oBbYlvaDamAK0EAA8AEI7Vj7YjvOB6Rxwb6-aboGAJJEAJnh90OzieyBATAY4Vo0EmICSjhgAHUcBcYORkAAOHxTJIQDYUnYJU7Ne7JACO6BwyVexA+XyQax+9w27xwXPOIEExxkAEV0BB4BipjBiAwqTS6UgACxanbSY5KCDvT64uCCACsJPQGzMOoW4ru-n8QA)
```sh
# Options (if any):
 prettier --single-quote --tab-width 4 --trailing-comma es5
```

**Input:**
```jsx
function watch(callback) {
    fs.watchFile(
        CONFIGFILEPATH,
        {interval: 1000},
        async (cur, prev) => {
            let config = await get(true)
            callback(config)
        }
    )
}
```

**Output:**
```jsx
function watch(callback) {
    fs.watchFile(CONFIGFILEPATH, { interval: 1000 }, async (cur, prev) => {
        let config = await get(true);
        callback(config);
    });
}
```

**Expected behavior:**
I don't want prettier to put some of the arguments in one line, and some of the arguments in extra lines.
Is there a way to disable this?

**Comment by josephfrazier** [comment]
If you're willing to break the object containing `interval` across multiple lines, prettier will preserve that and therefore put the other arguments on their own lines as well: [playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEA7gIYxgAWAFGMQDa0BGxYA1gJT7AA6U++qAZwB0JMuQBiOWnEo8+fAMIB5AHLiAkgHENAGQCiABQCCAFQASAGjnzuveXxyw4AJwBudJPgCMABj-W+AF8rOz5iAQBPLHxqdGcLfAAHZzhXDgBeAD5OAPlpGHxIKFQcAHN8dPxiEhwC0rgYShhndDg2AG5cvhp6JlZqaBLSjtzA6xGoQJALEAhE3GgBZFBkxxgAdRwAExhyZAAOHxmYYgZNnb2kACYZgUdS6QBFdAh4ZFQ6AThj52Ipe4UEAAtkDiMgQFBoHBpiAGL9WA0AMqJZj3ZDNVozABWAgAHgAheEsJHEIFwHSOaFID60L4zFHOL7OcFMBgRWjQGFfIE4dEtb4gdBfEynJbUz4CxxMmAGX6lUHvCUzFIAR3QOBSsuI8rB4tpAuSEC+61+iT5mJAWwgYEV+pm4Rgtrps3QMESrquTrggUCQA)

```js
function watch(callback) {
  fs.watchFile(
    CONFIGFILEPATH,
    {
      interval: 1000
    },
    async (cur, prev) => {
      let config = await get(true);
      callback(config);
    }
  );
}
```

**Comment by j-f1** [comment]
This is an intentional behavior to better support callback-style asynchronous programming by only indenting code in the callback once instead of twice. We generally don’t add options to support stylistic changes, so there isn’t a way to disable this.

**Comment by brabeji** [comment]
Hello @j-f1!

> to better support callback-style asynchronous programming by only indenting code in the callback once instead of twice.

Could you please give an example of what do you mean? *That* really seems like a stylistic issue. I understand that `prettier` is by design opinionated and if you guys use callback-style asynchronous programming in your projects, there's not much we can do.

BUT, following code is a mess. The indentation level in callback doesn't reflect the real code structure at all, MakingItHarderToRead™️.
```
fs.watchFile(CONFIGFILEPATH, { interval: 1000 }, async (cur, prev) => {
    let config = await get(true);
    callback(config);
});
```
I would like to decide where the newlines are myself.

**Comment by j-f1** [comment]
@brabeji That’s the case I was talking about. Since all the other parameters fit on one line, we keep them on one line instead of breaking them onto new lines and indenting the callback body another level.

> I would like to decide where the newlines are myself.

This is not something we want to support. Prettier’s goal is to strip away your formatting and replace it with its own, with few exceptions. Looking at whether you chose to break the arguments into multiple lines goes against that goal and reduces Prettier’s usefulness by making your code inconsistent — some of your team members may choose to format the function on one line while others will break it onto multiple lines, creating the inconsistency that Prettier tries to prevent.

**Comment by brabeji** [comment]
Thank you for getting back to me @j-f1.
>> I would like to decide where the newlines are myself.

Shouldn't have said that. I'm totally fine with prettier changing my

```
foo(
  {a: 'b'}
)
```
to
```
foo({ a: 'b' })
```

The issue lies in variable indentation of constantly nested code (if you will).

> Since all the other parameters fit on one line...

Well, *all* parameters don't fit on one line. That callback function taken as a whole is either too long on a single line or is broken up to multiple lines (it doesn't fit either way). *all the other parameters* being defined as all parameters which are not the only inline function at the end of parameter list seems too arbitrary. ([I believe that's how things work right now](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEA7gIYxgAWAFGMQDa0BGxYA1gJT7AA6U+f+qAM4A6EmXIAxHLTiUe-BfgDCAeQByEgJIBxLQBkAogAUAggBUAEgBp5ivsByw4AJwBudJPgCMABj8BfG147fGJBAE8sfGp0Zyt8AAdnOFcOAF4APk5bEP4Aejz8LhAYchxnABN8GVc4WnwoOEFcKABzYviAJmqUusF8CFR8RwqEGBzcnph8SChUHFb8NNCSHGnWuBhKGGd0ODYJ3Jp6JlZqaHnWg+C7fwnru6geDCxcAjEKajpGZnZsm4EIg+kmkskO-FUGh0+mM5gsAHVNJYAEoGEx6PQATT06m0ahMAFkDEFcg4nG4PN4-D5AuC+GFImBomBYvEkil0lluADcgUiiUypUerV6o1mo52iB4gBmYV9AZDEZjfAAWmG01GCQQFX6BBkbVKCvwmGStFIcCq7mcOGIDBkDWIAFs4HTFDJprNLksVsQ1vgNlsdntrpN+McfmdPQsQyFHgoHjxnphsHheMCvidfhxuQohKJSBQpDI5DyIRIjCSQmT4BTaJ5fAFK3YGVEYnFEslUksua6FHziqVylUanUGk0WpKZXLaLrFVBRrBVer8JrtbreOzBAgwHAjSa6uaqgliM4nZsXL3+O6ZhcFt7iKt1pttrt9pe+OHTixznNo67aaWXAwAAGvcPCPFKIAQAkbxQIIyCgCezgQIQRgngg8EoMQrgQDgFSQQwp6sJsADKx5gBKyBBnAVggI4W7ODARinq0jrEMgqB0FutEAFaCAAHgAQkRLCkWeeiOHAHFcTRIDHs4DHICATAMOEtDQJBSSODA8J4aUyAABw+LRSQQFu8KngkVGvrRyQAI7oOUcDMcQrHsUgnEzrJW6Ojg1l7LRggSjIACK6AQPA0lebRMC2rpFT6UgAAsMWntIEpKBAjpsUpTQAKyQegW5mLamGedxIBhDAUUVRUEBgDVskQOgMAJC1nSNf4-hAA))

Why not format:
```
fs.watchFile(async (cur, prev) => {
    let config = await get(true);
    callback(config);
});
```
as
```
fs.watchFile(
    async (cur, prev) => {
        let config = await get(true);
        callback(config);
    }
);
```
In other words: Parameters either all fit on one line or are on separate lines each. Thats something I would love to be able to configure.

I hope that any of above makes sense to you. Thank you very much!

**Comment by j-f1** [comment]
@brabeji the reason we don’t format it that way (or at least _my_ reasoning) is that the callback pattern is fairly common, and so are nested callbacks, so we format it that way to reduce the amount the code in the callback is indented. For example, in your preferred example, the code is indented 8 spaces instead of 4, potentially reducing the callback’s legibility, especially if callbacks are nested.

**Comment by Dzenly** [comment]
I also wish to have an option to disable this arguments mess when user already separated arguments by lines, because it looks ugly and creates unexpected git conflicts.

**Comment by bypotatoes** [comment]
Totaly agree with topic starter, at least we should do something with short arrow functions cause what we have now:
```
const getUserResources = createSelector(
  getResources,
  resources => resources.filter(r => r.type === 'USER'),
);
```
converts into 
```
const getHrResources = createSelector(getResources, resources =>
  resources.filter(r => r.category === "HR")
);
```
